llvm/include/llvm/CAS/CASDB.h

```diff
@@ -12,170 +12,414 @@
+///     - \a storeNode(), \a storeBlob(), and \a storeTree() are "get-or-create"
+///       methods, writing an object if it doesn't exist yet, and return a
+///       handle to it in any case.
+///     - \a loadObject(const CASID&) looks up an object by its UID.
+/// - Objects can reference other objects, forming an arbitrary DAG.
+///
+/// There are currently three kinds of objects.
+///
+/// - ObjectKind::Node: Contains 0+ bytes of data and a list of 0+ \a
+///   Reference. A \a Reference either points at another CAS object (\a
+///   ReferenceKind::Object) or at raw data (\a ReferenceKind::RawData).
+///   Created with \a storeNode().
+/// - ObjectKind::Tree: Sorted list of 0+ \a NamedTreeEntry, which associates a
+///   name, another CAS object, and a \a TreeEntry::EntryKind. Designed to
+///   represent a filesystem in the CAS. Created with \a storeTree().
+/// - ObjectKind::Blob: Contains 0+ bytes of data. Designed to represent an
+///   opaque file. Created with \a storeBlob().
+///
+/// TODO: Remove Blob and Tree, moving these concepts to a filesystem schema
+/// that sits on top of the CAS.
+///
+/// \a storeData() stores raw data of 0+ bytes for uniquing without requesting
+/// a UID. It's appropriate when a UID will not be needed because it will only
+/// be referenced directly (or by content), never by \a CASID.
+///
+/// - A CAS schema is free to fall back to blobs or leaf nodes for all raw data
+///   requests (default), or for all requests whose content matches a
+///   particular pattern (e.g., ends with null, bigger than a particular size).
+/// - The CAS schema's choice may leak into the parent context. \a storeData()
+///   returns \a AnyDataHandle, which could be raw data or a CAS object handle.
+/// - CAS implementations must consistent. E.g., \c storeData("D1") should have
+///   a consistent return across the schema (either \a ReferenceKind::Object or
+///   \a ReferenceKind::RawData), regardless of whether \c storeNode(None,"D1")
+///   has been called at some point. Further, the UID of a node N1 that
+///   references that raw data "D1" must be consistent.
+/// - Conceptually, raw data is "part of" any object that references. Raw data
+///   must be serialized and hashed as part of the parent object, since it can
+///   only be referenced by content (it has no UID). As such, it's implicitly
+///   loaded any time an object that directly references it is loaded.
+///
+/// The \a CASDB interface has a few ways of referencing objects and raw data
+/// in the CAS.
+///
+/// - \a Reference: a reference to an object or raw data in this CAS instance,
+///   obtained through: \a CASDB::readRef() or \a CASDB::forEachRef(), which
+///   are accessors for node CAS objects; or \a CASDB::getReference(), which
+///   looks up a \a CASID without loading the object. A \a Reference points at
+///   a node that is known to exist in this CAS instance, but it may not be
+///   loaded (in a distributed CAS, it may not even be local!).
```

Comment:
What does it mean to be "known to exist in this CAS instance" for something other than the in-memory CAS? I can think of two interpretations
* It existed at some point in the past (ie. at the time it gave you the Reference)
* It is pinned, so it cannot be garbage collected while this CASDB is alive.

Reply:
I think, both.

This is as opposed to `CASID`, which may have never existed in this CAS instance; a `CASID` could have been from a different on-disk CAS with the same schema, just happens to have the same schema.

Reply:
I meant the former:

- A `Reference` is something that this CAS knows about.
- This is unlike `CASID`, which has never been guaranteed to be something that this CAS knows about.

But the latter is also true, I think:

- Any object accessed/created during the lifetime of this `CASDB` instance should be protected from garbage collection as long as this `CASDB` is around. So, any object transitively referenced by this `CASDB` instance should be safe too.